### PR TITLE
Laboratory work #3, Manukyan Veronica - 20FPL2

### PR DESCRIPTION
**Pull request by veronicamanukyan**

### CODE REVIEW

File: lab_3/main.py [hunk 1]
```diff
@@ -20,8 +21,41 @@ def tokenize_by_sentence(text: str) -> tuple:
...
+        elif letter == 'ß':
+            text = text.replace(letter, 'ss')
+    regexp = re.compile('[.!?] ?')
+    sentences = re.split(regexp, text)
+    cleaned_sentences = []
+    for sentence in sentences:
+        cleaned_sentence = sentence.strip()
+        if cleaned_sentence:
+            cleaned_sentences.append([cleaned_sentence])
+    for ind, sentence in enumerate(cleaned_sentences):
```
Comment (dmitry-uraev) [inline lab_3/main.py:48]: ideally, you can replace all these for loops with just one for loop

File: lab_3/main.py [hunk 2]
```diff
@@ -31,38 +65,61 @@ class LetterStorage:
...
         """
         Puts a letter into storage, assigns a unique id
         :param letter: a letter
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if not isinstance(letter, str):
+            return -1
+        if letter not in self.storage:
+            self.count += 1
```
Comment (dmitry-uraev) [inline lab_3/main.py:79]: you can add ids even without self.count parameter. Guess how?

File: lab_3/main.py [hunk 3]
```diff
@@ -73,7 +130,12 @@ def encode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
     :param corpus: a tuple of sentences
     :return: a tuple of the encoded sentences
     """
-    pass
+    if not isinstance(storage, LetterStorage) or not isinstance(corpus, tuple):
+        return()
+    encoded_corpus = tuple(
+                     tuple(tuple(storage.get_id_by_letter(number) for number in word)
+                           for word in sentence) for sentence in corpus)
+    return encoded_corpus
```
Comment (dmitry-uraev) [inline lab_3/main.py:138]: good

File: lab_3/main.py [hunk 4]
```diff
@@ -84,7 +146,12 @@ def decode_corpus(storage: LetterStorage, corpus: tuple) -> tuple:
     :param corpus: an encoded tuple of sentences
     :return: a tuple of the decoded sentences
     """
-    pass
+    if not isinstance(storage, LetterStorage) or not isinstance(corpus, tuple):
+        return()
+    decoded_corpus = tuple(
+                     tuple(tuple(storage.get_letter_by_id(number) for number in word)
+                           for word in sentence) for sentence in corpus)
+    return decoded_corpus
```
Comment (dmitry-uraev) [inline lab_3/main.py:154]: good

File: lab_3/main.py [hunk 5]
```diff
@@ -116,7 +186,25 @@ def extract_n_grams(self, encoded_corpus: tuple) -> int:
...
+                last_symbol = first_symbol + self.size
+                while last_symbol <= len(word):
+                    n_gram = word[first_symbol: last_symbol]
+                    first_symbol += 1
+                    last_symbol += 1
+                    words.append(tuple(n_gram))
+                sentences.append(tuple(words))
+            corpus.append(tuple(sentences))
+        self.n_grams = tuple(corpus)
+        return 0
```
Comment (dmitry-uraev) [inline lab_3/main.py:207]: does this algorithm allows you to create n-grams? e.g. 1-grams and 4-grams?

File: lab_3/main.py [hunk 6]
```diff
@@ -187,7 +287,16 @@ def create_from_tokens(self, encoded_corpus: tuple, ngram_sizes: tuple) -> int:
             (((1, 2), (2, 3), (3, 1)), ((1, 4), (4, 5), (5, 1)), ((1, 2), (2, 6), (6, 7), (7, 7), (7, 8), (8, 1))),
         )
         """
-        pass
+        if not isinstance(encoded_corpus, tuple) or not isinstance(ngram_sizes, tuple):
+            return 1
+        for size in ngram_sizes:
```
Comment (dmitry-uraev) [inline lab_3/main.py:292]: you can rewrite this method to use only one for loop

File: lab_3/main.py [hunk 7]
```diff
@@ -213,7 +322,18 @@ def get_top_k_n_grams(self, k: int, trie_level: int) -> tuple:
...
+        if not isinstance(k, int) or not isinstance(trie_level, int):
+            return ()
+        if trie_level < 1 or k < 1:
+            return ()
+        for trie in self.tries:
+            if trie.size == trie_level:
+                sorted_keys = tuple(sorted(trie.n_gram_frequencies,
+                                           key = trie.n_gram_frequencies.get, reverse = True)[:k])
+                return sorted_keys
+        return()
```
Comment (dmitry-uraev) [inline lab_3/main.py:334]: good

File: lab_3/main.py [hunk 8]
```diff
@@ -253,7 +373,20 @@ def calculate_distance(unknwon_profile: LanguageProfile, known_profile: Language
...
+    distance = 0
+    top_un = unknown_profile.get_top_k_n_grams(k, trie_level)
+    top_kn = known_profile.get_top_k_n_grams(k, trie_level)
+    for ind_un, element_un in enumerate(top_un):
+        if element_un not in top_kn:
+            distance += len(top_kn)
+        for ind_kn, element_kn in enumerate(top_kn):
+            if element_kn == element_un:
+                distance += abs(ind_un - ind_kn)
+    return distance
```
Comment (dmitry-uraev) [inline lab_3/main.py:389]: good

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
Good start. Let's go for 8-10 now?

File: lab_3/main.py [hunk 9]
```diff
@@ -20,8 +21,41 @@ def tokenize_by_sentence(text: str) -> tuple:
...
     """
-    pass
-
+    if not isinstance(text, str):
+        return ()
+    invaluable_trash = ['`', '~', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '-', '+',
+                        '=', '{', '[', ']', '}', '|', '\\', ':', ';', '"', "'", '<', ',', '>',
+                        '/', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0']
+    text = text.lower()
+    for symbol in invaluable_trash:
```
[the remark this answers is not in the document]
Comment (veronicamanukyan) [reply lab_3/main.py:30]: I used dictionary

File: lab_3/main.py (hunk 5 above)
Comment (veronicamanukyan) [reply lab_3/main.py:196]: guess, now it does)

File: lab_3/main.py (hunk 1 above)
Comment (veronicamanukyan) [reply lab_3/main.py:42]: is it necessary to change it or can I leave it as it is?

**Pull request closed without merging**